### DISCUSSION TITLE
Model RAG every answer shows doc even when not relevant

### DISCUSSION THREAD

**Discussion by stevenbaert** [open-webui/open-webui#4823] [Q&A] [answered]
Hope this screenshot makes it clear. Why does a model always shows the PDF link even if it is not relevant?
Thanks for your feedback. Let me know if unclear! 

![image](https://github.com/user-attachments/assets/bb3494e5-5a6e-4eb2-b1f9-0f00e285237f)

**Comment by PlebeiusGaragicus** [comment]
I have huge skepticism about OI's RAG implementation right now.

What I'd like to see is OI become a frontend for maintaining collections of documents, which can then be embedded as needed and available for LLMs (agents) to pull from.  What I'm not confident about is this naive approach to RAG.  Modern techniques (and this space moves very very quickly) use more of a "flow engineering" approach as retrieved documents are re-ranked and graded (being excluded if found to be irrelevant) before being passed to an LLM where more steps can be taken before final generation.

If OI exposes an API for its "document" vector store / RAG thing - that'd be greeaaaat.  Then my agent flows can just use it and I can use OI as my document management frontend as well as my chat interface.

I think this project is just too young to have it built out yet, but I think you're correct, the current system will often have a model try to do retrieval even if completely unnecessary.  Similar with tool use - if you provide it to the LLM it may just pull the trigger.

Relevant papers: [RQ-RAG](https://arxiv.org/pdf/2404.00610), [Corrective RAG](https://arxiv.org/pdf/2401.15884v2), [Self-RAG](https://arxiv.org/pdf/2310.11511v1)

**Comment by stevenbaert** [comment]
What stands out to me on this forum is the lack of responses. When there is a response, after days or weeks, it often doesn’t address the question and instead goes off-topic. While I appreciate the effort, it doesn’t help.

**Comment by justinh-rahb** [comment, marked as answer]
When including a document by drag-dropping, or `#` shortcut, you can remove it from context in future chat turns:

<img width="384" alt="Screenshot 2024-08-27 at 3 31 15 AM" src="https://github.com/user-attachments/assets/56beb9f2-f82e-4415-8115-a0ea236df44a">

If it's been set in the model config, then it's going to be referenced in all, as designed.

**Comment by stevenbaert** [comment]
I’m not sure if I understand. How can you offer documents in a proper way? 

You just uploading the documents, references then should only be shown when a RAG process occurs, and not shown when there’s nothing to retrieve, right? 

It can’t be that you need to manually tag or upload the document then remove it afterwards(?)

What’s the reasoning behind showing the reference in every call when the document is already referenced in the model configuration? You mentioned this is ‘by design,’ but it doesn’t seem logical to me.

**Reply by PlebeiusGaragicus** [reply]
After reviewing the current RAG feature it seems unreliable - does that help?  It's an immature feature yet to be fully developed.

If you need RAG that works - you may have to built it yourself.  That was my earlier point.

**Reply by stevenbaert** [reply]
Pipelines seem to be the answer for RAG, although it's not documented anywhere. At least, I haven’t found any documentation. 

I might document it myself later when I fully have it working as it should. There is quite a bit of documentation on Open-Webui, but on specifics it is lacking info. Then you smile turn to the forum, but there responses are slow, none or not to the point. Which is a petty since that would boost adoption more widely within the community.

**Comment by qiulang** [comment]
@stevenbaert @justinh-rahb Can you guys look into my question about RAG here https://github.com/open-webui/open-webui/discussions/5002 

I can't make my RAG work at all.
